cmd/block-generator/runner/run.go

```diff
@@ -82,31 +83,133 @@ func (r *Args) run() error {
 	return nil
 }
 
-// Helper to record the import rate.
-func recordDataToFile(entry Entry, key string, out *os.File) error {
+type metricType int
+
+const (
+	rate metricType = iota
+	intTotal
+	floatTotal
+)
+
+// Helper to record metrics. Supports rates (sum/count) and counters.
+func recordDataToFile(entry Entry, prefix string, out *os.File) error {
+	num := 0
+	var writeErrors strings.Builder
+	var writeErr error
+	record := func(name string, t metricType) {
+		key := fmt.Sprintf("%s_%s", prefix, name)
+		if err := recordMetricToFile(entry, out, key, name, t); err != nil {
+			writeErr = err
+			if writeErrors.Len() > 0 {
+				writeErrors.WriteString(", ")
+			}
+			writeErrors.WriteString(name)
+		}
+		num++
+	}
+
+	record(metrics.BlockImportTimeName, rate)
+	record(metrics.CumulativeImportTimeName, floatTotal)
+	record(metrics.ImportedTxnsPerBlockName, rate)
+	record(metrics.CumulativeTxnsName, intTotal)
+	record(metrics.CurrentRoundGaugeName, intTotal)
+	record(metrics.BlockUploadTimeName, rate)
+	record(metrics.CumulativeBlockUploadTimeName, floatTotal)
+
+	if writeErrors.Len() > 0 {
+		return fmt.Errorf("error writing metrics (%s): %w", writeErrors.String(), writeErr)
+	}
+	if num != len(metrics.AllMetricNames) {
+		return fmt.Errorf("only %d of %d metrics written", num, len(metrics.AllMetricNames))
+	}
+
+	// Calculate import transactions per second.
+	totalTxn, err := getMetric(entry, metrics.CumulativeTxnsName, intTotal)
+	if err != nil {
+		return err
+	}
+
+	importTimeS, err := getMetric(entry, metrics.CumulativeImportTimeName, intTotal)
+	if err != nil {
+		return err
+	}
+	tps := totalTxn / importTimeS
+	tpsKey := "overal_transactions_per_second"
+	msg := fmt.Sprintf("%s_%s:%.2f\n", prefix, tpsKey, tps)
+	if _, err := out.WriteString(msg); err != nil {
+		return fmt.Errorf("unable to write metric '%s': %w", tpsKey, err)
+	}
+
+	return nil
+}
+
+func recordMetricToFile(entry Entry, out *os.File, outputKey, metricSuffix string, t metricType) error {
+	value, err := getMetric(entry, metricSuffix, t)
+	if err != nil {
+		return err
+	}
+
+	var msg string
+	if t == intTotal {
+		msg = fmt.Sprintf("%s:%d\n", outputKey, uint64(value))
+	} else {
+		msg = fmt.Sprintf("%s:%.2f\n", outputKey, value)
+	}
+
+	if _, err := out.WriteString(msg); err != nil {
+		return fmt.Errorf("unable to write metric '%s': %w", outputKey, err)
+	}
+
+	return nil
+}
+
+func getMetric(entry Entry, suffix string, t metricType) (float64, error) {
+	total := 0.0
 	sum := 0.0
 	count := 0.0
+	hasSum := false
+	hasCount := false
 
 	for _, metric := range entry.Data {
```

Comment:
Will `entry.Data` contain same key multiple times? Does this loop select the latest measurement then?

Reply:
There is an array of entry objects and 10 samples are taken while the test runs.